File: fetch-deals.js
const fs = require('fs');

const RAPIDAPI_KEY = process.env.RAPIDAPI_KEY;
const RAPIDAPI_HOST = 'sky-scrapper.p.rapidapi.com';

if (!RAPIDAPI_KEY) {
  console.error('ERROR: RAPIDAPI_KEY secret is not set!');
  process.exit(1);
}

const ORIGINS = ['PHL', 'JFK', 'EWR', 'BWI', 'DCA', 'BOS'];

const REGIONS = {
  MIA: 'Southeast', ATL: 'Southeast', MCO: 'Southeast', TPA: 'Southeast',
  ORD: 'Midwest', DTW: 'Midwest', MSP: 'Midwest', CLE: 'Midwest',
  LAX: 'West', SFO: 'West', DEN: 'West', SEA: 'West', LAS: 'West',
  DFW: 'Southwest', PHX: 'Southwest', HOU: 'Southwest', AUS: 'Southwest',
  SJU: 'Southeast', CUN: 'Southwest', NAS: 'Southeast',
};

async function fetchDealsFromOrigin(origin, travelDate) {
  const url = `https://${RAPIDAPI_HOST}/api/v1/flights/searchFlightEverywhere?originSkyId=${origin}&oneWay=false&currency=USD&countryCode=US&market=en-US&travelDate=${travelDate}`;

  console.log(`Fetching from ${origin}...`);

  try {
    const response = await fetch(url, {
      headers: {
        'x-rapidapi-host': RAPIDAPI_HOST,
        'x-rapidapi-key': RAPIDAPI_KEY,
      },
    });

    if (!response.ok) {
      console.log(`  API error ${response.status}`);
      return [];
    }

    const json = await response.json();
    const results = json?.data?.results || json?.data?.everywhere || json?.data || [];

    if (!Array.isArray(results)) {
      console.log(`  Unexpected response format`);
      return [];
    }

    console.log(`  Found ${results.length} destinations`);

    return results.slice(0, 15).map(item => ({
      destination: item.destinationSkyId || item.skyId || item.iata || 'UNK',
      region: REGIONS[item.destinationSkyId] || REGIONS[item.skyId] || 'Other',
      price: item.price?.amount || item.cost || item.price || 0,
      airline: item.carrier || item.airline || 'Multiple',
      departTime: '',
      arriveTime: '',
      duration: '',
      stops: item.stops ?? 0,
      departDate: travelDate,
    })).filter(d => d.price > 0);

  } catch (e) {
    console.log(`  Error: ${e.message}`);
    return [];
  }
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function main() {
  const today = new Date();
  const travelDate = new Date(today.getTime() + 14 * 24 * 60 * 60 * 1000).toISOString().split('T')[0];

  console.log(`Fetching flight deals for ${travelDate}\n`);

  const result = {
    lastUpdated: new Date().toISOString(),
    departDate: travelDate,
    returnDate: new Date(today.getTime() + 18 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
    deals: {},
  };

  for (const origin of ORIGINS) {
    result.deals[origin] = await fetchDealsFromOrigin(origin, travelDate);
    await sleep(1000); // Rate limit
  }

  fs.writeFileSync('deals.json', JSON.stringify(result, null, 2));
  console.log('\nSaved deals.json');

  const totalDeals = Object.values(result.deals).reduce((sum, arr) => sum + arr.length, 0);
  console.log(`Total: ${totalDeals} deals`);
}

main().catch(e => {
  console.error('Fatal error:', e);
  process.exit(1);
});

File: fetch_deals.py
#!/usr/bin/env python3
"""Fetch flight deals using fli (Google Flights)."""

import json
from datetime import datetime, timedelta

from fli.search import SearchFlights
from fli.models import FlightSearchFilters, FlightSegment, PassengerInfo, SeatType, MaxStops, SortBy

ORIGINS = ["PHL", "JFK", "EWR", "BWI", "DCA", "BOS"]

DESTINATIONS = [
    ("MIA", "Southeast"),
    ("ATL", "Southeast"),
    ("ORD", "Midwest"),
    ("DEN", "West"),
    ("LAX", "West"),
    ("SFO", "West"),
    ("PHX", "Southwest"),
    ("DFW", "Southwest"),
]


def get_travel_dates():
    """Get departure and return dates (2 weeks out, 4-day trip)."""
    depart = datetime.now() + timedelta(days=14)
    ret = depart + timedelta(days=4)
    return depart.strftime("%Y-%m-%d"), ret.strftime("%Y-%m-%d")


def fetch_deals_for_origin(origin: str, depart_date: str, return_date: str) -> list:
    """Fetch flight deals from a single origin."""
    deals = []
    search = SearchFlights()

    for dest_code, region in DESTINATIONS:
        try:
            filters = FlightSearchFilters(
                passenger_info=PassengerInfo(adults=1),
                flight_segments=[
                    FlightSegment(
                        departure_airport=origin,
                        arrival_airport=dest_code,
                        travel_date=depart_date,
                    ),
                    FlightSegment(
                        departure_airport=dest_code,
                        arrival_airport=origin,
                        travel_date=return_date,
                    ),
                ],
                seat_type=SeatType.ECONOMY,
                stops=MaxStops.ANY,
                sort_by=SortBy.CHEAPEST,
            )

            results = search.search(filters)

            for flight in results[:3]:
                # Extract first leg info
                first_leg = flight.legs[0] if flight.legs else None

                deals.append({
                    "destination": dest_code,
                    "region": region,
                    "price": flight.price,
                    "airline": first_leg.airline if first_leg else "Unknown",
                    "departTime": first_leg.departure_time if first_leg else "",
                    "arriveTime": first_leg.arrival_time if first_leg else "",
                    "duration": f"{flight.duration // 60}h {flight.duration % 60}m",
                    "stops": flight.stops,
                    "departDate": depart_date,
                    "returnDate": return_date,
                })
        except Exception as e:
            print(f"Error fetching {origin} -> {dest_code}: {e}")
            continue

    # Sort by price and return top deals
    deals.sort(key=lambda x: x["price"])
    return deals[:10]


def main():
    depart_date, return_date = get_travel_dates()

    all_deals = {}
    for origin in ORIGINS:
        print(f"Fetching deals from {origin}...")
        all_deals[origin] = fetch_deals_for_origin(origin, depart_date, return_date)

    output = {
        "lastUpdated": datetime.utcnow().isoformat() + "Z",
        "departDate": depart_date,
        "returnDate": return_date,
        "deals": all_deals,
    }

    with open("deals.json", "w") as f:
        json.dump(output, f, indent=2)

    print("Done! Saved to deals.json")


if __name__ == "__main__":
    main()
